### PR TITLE
ENH: add from_dummies

### LINKED ISSUE #8745: API/ENH: from_dummies
Motivating from [SO](http://stackoverflow.com/questions/26762100/reconstruct-a-categorical-variable-from-dummies-in-pandas)

This is the inverse of `pd.get_dummies`. So maybe `invert_dummies` is better?
I think this name makes more sense though. 

This seems a reasonable way to do it. Am I missing anything?

```
In [46]: s = Series(list('aaabbbccddefgh')).astype('category')

In [47]: s
Out[47]: 
0     a
1     a
2     a
3     b
4     b
5     b
6     c
7     c
8     d
9     d
10    e
11    f
12    g
13    h
dtype: category
Categories (8, object): [a < b < c < d < e < f < g < h]

In [48]: df = pd.get_dummies(s)

In [49]: df
Out[49]: 
    a  b  c  d  e  f  g  h
0   1  0  0  0  0  0  0  0
1   1  0  0  0  0  0  0  0
2   1  0  0  0  0  0  0  0
3   0  1  0  0  0  0  0  0
4   0  1  0  0  0  0  0  0
5   0  1  0  0  0  0  0  0
6   0  0  1  0  0  0  0  0
7   0  0  1  0  0  0  0  0
8   0  0  0  1  0  0  0  0
9   0  0  0  1  0  0  0  0
10  0  0  0  0  1  0  0  0
11  0  0  0  0  0  1  0  0
12  0  0  0  0  0  0  1  0
13  0  0  0  0  0  0  0  1

In [50]: x = df.stack()

# I don't think you actually need to specify ALL of the categories here, as by definition
# they are in the dummy matrix to start (and hence the column index)
In [51]: Series(pd.Categorical(x[x!=0].index.get_level_values(1)))
Out[51]: 
0     a
1     a
2     a
3     b
4     b
5     b
6     c
7     c
8     d
9     d
10    e
11    f
12    g
13    h
Name: level_1, dtype: category
Categories (8, object): [a < b < c < d < e < f < g < h]
```

NB. this is buggy ATM.

```
In [51]: Series(pd.Categorical(x[x!=0].index.get_level_values(1)),categories=df.categories)
```

### PR DESCRIPTION
**Pull request by MarcoGorelli**
Just a first draft

Will close #8745 

Screenshot of examples

image

### CODE REVIEW

File: pandas/core/reshape/reshape.py [hunk 1]
```diff
@@ -751,6 +751,138 @@ def _convert_level_number(level_num, columns):
...
+    ----------
+    data : DataFrame
+    columns : list-like, default None
+        Column names in the DataFrame to be decoded.
+        If `columns` is None then all the columns will be converted.
+    prefix_sep : str, default '_'
+        Separator between original column name and dummy variable
+    dtype : dtype, default 'category'
+        Data dtype for new columns - only a single data type is allowed
+    fill_first : str, list, or dict, default None
```
Comment (WillAyd) [inline pandas/core/reshape/reshape.py]: I think I'm -1 on this feature; not sure I see it being super useful outside of niche applications

File: pandas/core/reshape/reshape.py [hunk 2]
```diff
@@ -751,6 +751,138 @@ def _convert_level_number(level_num, columns):
...
 
 
+def from_dummies(data, columns=None, prefix_sep="_", dtype="category", fill_first=None):
+    """
+    The inverse transformation of ``pandas.get_dummies``.
+
+    Parameters
+    ----------
+    data : DataFrame
+    columns : list-like, default None
```
Comment (WillAyd) [inline pandas/core/reshape/reshape.py]: I would also rather not provide `columns` as a keyword; duplicative of the first argument

Comment (WillAyd) [reply pandas/core/reshape/reshape.py]: Or is this supposed to be the inverse of the `prefix` argument in `get_dummies`? If so, I think should just reuse that name

**Review by WillAyd — CHANGES_REQUESTED** [review]
Cool thanks! Initial comments on API decisions

File: pandas/core/reshape/reshape.py (hunk 1 above)
Comment (MarcoGorelli) [reply pandas/core/reshape/reshape.py]: @WillAyd thanks for your feedback - when you say "this feature", do you mean `from_dummies`, or just `fill_first`?

File: pandas/core/reshape/reshape.py (hunk 2 above)
Comment (MarcoGorelli) [reply pandas/core/reshape/reshape.py]: Sure, I'll change it to `prefix`

File: pandas/core/reshape/reshape.py (hunk 1 above)
Comment (WillAyd) [reply pandas/core/reshape/reshape.py]: fill_first

**Comment by MarcoGorelli** [comment]
@WillAyd have made some updates. Have removed `fill_first`. Haven't added equivalents of `dummy_na` or `sparse` - would you prefer to get a bare-bones implementation out first and add those in future PRs, or should I add them now?

File: pandas/core/reshape/reshape.py [hunk 3]
```diff
@@ -751,6 +751,159 @@ def _convert_level_number(level_num, columns):
     return result
 
 
+def from_dummies(data, prefix=None, prefix_sep="_", dtype="category") -> "DataFrame":
```
Comment (WillAyd) [inline pandas/core/reshape/reshape.py:754]: Can you annotate these?

File: pandas/core/reshape/reshape.py [hunk 4]
```diff
@@ -751,6 +751,159 @@ def _convert_level_number(level_num, columns):
...
+            if i in seen:
+                continue
+            seen.add(i)
+            prefix.append(i)
+    elif isinstance(prefix, str):
+        prefix = [prefix]
+
+    # Check each row sums to 1 or 0
+    def _validate_values(data):
+        if not all(i in [0, 1] for i in data.sum(axis=1).unique().tolist()):
```
Comment (WillAyd) [inline pandas/core/reshape/reshape.py:866]: IIUC can simplify this by doing a set difference of `data.sum(axis=1).unique()` and `{0, 1}`

File: pandas/tests/reshape/test_from_dummies.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+import pandas as pd
+import pandas._testing as tm
+
+
+@pytest.mark.parametrize(
+    "dtype, expected_dict",
+    [
+        ("str", {"col1": ["a", "a", "b"]}),
+        (str, {"col1": ["a", "a", "b"]},),
+        (None, {"col1": ["a", "a", "b"]}),
```
Comment (WillAyd) [inline pandas/tests/reshape/test_from_dummies.py:12]: Why `None` here instead of `"category"`?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/reshape/reshape.py (hunk 4 above)
Comment (MarcoGorelli) [reply pandas/core/reshape/reshape.py]: Indeed, if we're not supporting `fill_first`, we can just check that all values equal 1. Thanks!

**Comment by MarcoGorelli** [comment]
Note to self: this change from `ba72b59b44f15076877e784e16cbc66aa03e0adc` breaks the tests

```diff
-    # groupby dtype.name to collect e.g. Categorical columns
-    counts = data.dtypes.value_counts().groupby(lambda x: x.name).sum()
+    counts = data._data.get_dtype_counts()
```

**Comment by MarcoGorelli** [comment]
@WillAyd thanks for your review, have updated

**Comment by jreback** [comment]
@MarcoGorelli did you see this impl: https://github.com/pandas-dev/pandas/issues/8745#issuecomment-630924929.

I also don't see why we would ever return anyting but a Series / Categorical here. can you elaborate?

**Comment by MarcoGorelli** [comment]
> @MarcoGorelli did you see this impl: [#8745 (comment)](https://github.com/pandas-dev/pandas/issues/8745#issuecomment-630924929).
> 
> I also don't see why we would ever return anyting but a Series / Categorical here. can you elaborate?

I hadn't, thanks for pointing me towards it! @clbarnes that looks good, are you interested in taking over this issue with your implementation?

**Comment by clbarnes** [comment]
I can do if you'd prefer that strategy - will need to have a think about the requirements, e.g. what kind of incorrect inputs should it try to gracefully handle, handling masked arrays and NA and so on. That's probably discussion for the issue though.

I personally err on the side of being less permissive than pandas is generally so it may end up being pretty constrained, if that's OK with the maintainers.

**Comment by MarcoGorelli** [comment]
> I can do if you'd prefer that strategy

Sure, go ahead! I'll close this PR then, though please let me know if for whatever reason you chose not to work on this

**Pull request closed without merging**